src/Validation/Validator.php

```diff
@@ -0,0 +1,229 @@
+
+        if (isset($this->requirements)) {
+            // loop through shipmentoptions and merge the schema's obtained based on the setting
+            $this->adjustSchemaToRequirements();
+        }
+    }
+
+    /**
+     * @param  array $shipmentOptions
+     *
+     * @return array
+     */
+    private function convertNullValuesInShipmentOptions(array $shipmentOptions): array
+    {
+        foreach ($shipmentOptions as $shipmentOption => $value) {
+            if ('insurance' === $shipmentOption) {
+                $shipmentOptions[$shipmentOption] = (int) $value;
+                continue;
+            }
+
+            if ('labelDescription' === $shipmentOption) {
+                $shipmentOptions[$shipmentOption] = (string) $value;
+                continue;
+            }
+
+            $shipmentOptions[$shipmentOption] = (bool) $value;
+        }
+
+        return $shipmentOptions;
+    }
+
+    /**
+     * @param  string $what
+     * @param  string $column
+     * @param         $value
+     *
+     * @return $this
+     */
+    private function mergeIntoSchema(string $what, string $column, $value): self
+    {
+        if (! isset($this->extras[$what])) {
+            return $this;
+        }
+
+        $extras = $this->extras[$what];
+        $index  = array_search($value, array_column($extras, $column), true) ?: 0;
+        $extras = $extras[$index];
+
+        if (isset($extras['schema'])) {
+            $this->schema = $this->mergeSchemasRecursively($this->schema, $extras['schema']);
+        }
+
+        // todo implement json-schema draft 2019-09 or higher to integrate requirements into the schema validation
+        if (isset($extras['requirements'])) {
+            $this->requirements = $this->mergeSchemasRecursively($this->requirements, $extras['requirements']);
+        }
+
+        $this->extras = $extras;
+
+        return $this;
+    }
+
+    /**
+     * @param  array $schema
+     * @param  array $schema2
+     *
+     * @return array
+     */
+    private function mergeSchemasRecursively(array $schema, array $schema2): array
+    {
+        foreach ($schema2 as $key => $value) {
+            if (! isset($schema[$key]) || ! is_array($schema[$key])) {
+                $schema[$key] = [];
+            }
+
+            if (is_array($value) && 'enum' !== $key) {
+                $value = $this->mergeSchemasRecursively($schema[$key], $value);
+            }
+
+            $schema[$key] = $value;
+        }
+
+        return $schema;
+    }
+
+    /**
+     * @return array
+     */
+    public function getValidationSchema(): array
+    {
+        return $this->schema;
+    }
+
+    /**
+     * @return bool
+     * @throws \MyParcelNL\Pdk\Base\Exception\InvalidCastException
+     */
+    public function validate(): bool
+    {
+        $orderArray = $this->order->toArray();
+
+        $orderArray['deliveryOptions']['shipmentOptions'] = $this->convertNullValuesInShipmentOptions(
```

Comment:
waarom als array? Je hebt nu een extra functie, moet daarin convertNullValuesInShipmentOptions allemaal logica die niks met deze class te maken heeft vastleggen en een loop doen.

je kunt ook zoiets doen: 

```php
    $shipmentOptions->fill([
        'ageCheck'        => (bool) $shipmentOptions->ageCheck,
        'insurance'       => (int) $shipmentOptions->insurance,
        'largeFormat'     => (bool) $shipmentOptions->largeFormat,
        'onlyRecipient'   => (bool) $shipmentOptions->onlyRecipient,
        'return'          => (bool) $shipmentOptions->return,
        'sameDayDelivery' => (bool) $shipmentOptions->sameDayDelivery,
        'signature'       => (bool) $shipmentOptions->signature,
    ]);
```
of nog beter, voeg een method toe aan de ShipmentOptions class die dit doet. misschien willen we wel vaker alle waarden "vastleggen"